## Question 1 (iv)

In [4]:
# write a function (divisible_multiple) to count and display numbers between 152 and 1176 that are divisible bby 5 and are a multiple of 4
def divisible_multiple():
    for i in range(152,1176):
        if i % 5 == 0 and i % 4 == 0:
            print(i)

In [ ]:
divisible_multiple()

## Question 2

2a) The data structure is a dictionary. It has 2 keys. These keys are Student A and Student B. Each key-value pair maps the key to its associated value.


In [6]:
students = {'Student A': {'Math': 67, 'English': 78},'Student B': {'Math': 70, 'English': 70}}
print(students)

# list the keys in the dictionary
print(students.keys())
# list the values in the dictionary
print(students.values())

{'Student A': {'Math': 67, 'English': 78}, 'Student B': {'Math': 70, 'English': 70}}
dict_keys(['Student A', 'Student B'])


In [8]:
# access the English grade for student B
print(students['Student B']['English'])

70


In [10]:
# Change the English grade for student A to 90
students['Student A']['English'] = 90
print(students)

{'Student A': {'Math': 67, 'English': 90}, 'Student B': {'Math': 70, 'English': 70}}


In [11]:
# Add 'Student C' with Math grade of 80 and English grade of 99
students['Student C'] = {'Math': 80, 'English': 99}
print(students)

{'Student A': {'Math': 67, 'English': 90}, 'Student B': {'Math': 70, 'English': 70}, 'Student C': {'Math': 80, 'English': 99}}


In [ ]:
# Function to check if a key exists in a dictionary
def check_key(key):
    if key in students.keys():
        print("Key exists")
    else:
        print("Key does not exist")

The big O complexity if the key does not exist is O(n)

In [12]:
# from the students dictionary, write a function that returns the total for math and english for each student 
# and also return the average of the students for both math and english

def total_average():
    for key, value in students.items():
        print(key)
        print(value)
        print("Total: ", value['Math'] + value['English'])
        print("Average: ", (value['Math'] + value['English'])/2)

total_average()


Student A
{'Math': 67, 'English': 90}
Total:  157
Average:  78.5
Student B
{'Math': 70, 'English': 70}
Total:  140
Average:  70.0
Student C
{'Math': 80, 'English': 99}
Total:  179
Average:  89.5


## Question 4

In [15]:
!pip install matplotlib
!pip install seaborn

In [4]:
# Load libraries
import pandas as pd
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns


In [6]:
# Load dataset
hospital = pd.read_csv(r"C:\Users\Rick-Royal\Documents\Strathmore University Data Science and Analytics\Module 4\Module-IV\FCC\kenya-hospital-ke.csv")

In [7]:
# View first 5 rows of dataset
hospital.head()

Code                                  Name  \
0  25720                      Itete Dispensary   
1  25722  Penda Medical Center - Mathare North   
2   None                            Mukavakava   
3   None                    Nchurra Dispensary   
4   None             Upland Health Care Center   

                           Officialname Registration_number Keph level  \
0                      Itete Dispensary               01245    Level 2   
1  Penda Medical Center - Mathare North              017345    Level 3   
2                            Mukavakava                None    Level 2   
3                    Nchurra Dispensary                None    Level 2   
4             Upland Health Care Center                None    Level 2   

    Facility type Facility_type_category  \
0      Dispensary             DISPENSARY   
1  Medical Center         MEDICAL CENTER   
2      Dispensary             DISPENSARY   
3      Dispensary             DISPENSARY   
4  Medical Clinic         MEDICAL CLINIC   

                                  Owner          Owner type  \
0                    Ministry of Health  Ministry of Health   
1    Private Practice - Private Company    Private Practice   
2                    Ministry of Health  Ministry of Health   
3                    Ministry of Health  Ministry of Health   
4  Private Practice - Nurse / Midwifery    Private Practice   

      Regulatory body  ...            Ward  Operation status  Open_whole_day  \
0  Ministry of Health  ...         Koyonzo       Operational              No   
1          Kenya MPDB  ...   Mathare North       Operational              No   
2  Ministry of Health  ...  Butali/Chegulo       Operational              No   
3  Ministry of Health  ...          Melelo       Operational              No   
4                None  ...        Kerugoya       Operational              No   

  Open_public_holidays Open_weekends Open_late_night Service_names Approved  \
0                   No            No              No           NaN      Yes   
1                  Yes           Yes              No           NaN      Yes   
2                   No            No              No           NaN      Yes   
3                   No            No              No           NaN      Yes   
4                  Yes           Yes             Yes           NaN      Yes   

  Public visible Closed  
0            Yes     No  
1            Yes     No  
2            Yes     No  
3            Yes     No  
4            Yes     No  

[5 rows x 26 columns]

In [8]:
hospital.tail()

Code                             Name  \
12389  13311             Bour-Algy Dispensary   
12390  12547  Mtito andei Sub County Hospital   
12391  12402             Kyaani Health Centre   
12392  10834                Ndivai Dispensary   
12393  10907       Oaklands Estate Dispensary   

                          Officialname Registration_number Keph level  \
12389             Bour-algy Dispensary                None    Level 2   
12390  Mtito Andei Sub County Hospital                None    Level 4   
12391             Kyaani Health Centre                None    Level 3   
12392                Ndivai Dispensary                None    Level 2   
12393        Oakland Estate Dispensary                None    Level 2   

                Facility type Facility_type_category  \
12389              Dispensary             DISPENSARY   
12390  Primary care hospitals              HOSPITALS   
12391     Basic Health Centre          HEALTH CENTRE   
12392              Dispensary             DISPENSARY   
12393              Dispensary             DISPENSARY   

                                    Owner          Owner type  \
12389                  Ministry of Health  Ministry of Health   
12390                  Ministry of Health  Ministry of Health   
12391                  Ministry of Health  Ministry of Health   
12392                  Ministry of Health  Ministry of Health   
12393  Private Practice - Private Company    Private Practice   

          Regulatory body  ...          Ward  Operation status  \
12389  Ministry of Health  ...         Iftin       Operational   
12390  Ministry of Health  ...   Mtito Andei       Operational   
12391  Ministry of Health  ...        Thange       Operational   
12392  Ministry of Health  ...  Leshau/Pondo       Operational   
12393          Kenya MPDB  ...       Mwihoko       Operational   

       Open_whole_day Open_public_holidays Open_weekends Open_late_night  \
12389              No                   No            No              No   
12390             Yes                   No            No              No   
12391              No                   No            No              No   
12392              No                   No            No              No   
12393              No                   No            No              No   

      Service_names Approved Public visible Closed  
12389           NaN      Yes            Yes     No  
12390           NaN      Yes            Yes     No  
12391           NaN      Yes            Yes     No  
12392           NaN      Yes            Yes     No  
12393           NaN      Yes            Yes     No  

[5 rows x 26 columns]

In [9]:
hospital.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 12394 entries, 0 to 12393
Data columns (total 26 columns):
 #   Column                  Non-Null Count  Dtype  
---  ------                  --------------  -----  
 0   Code                    12394 non-null  object 
 1   Name                    12394 non-null  object 
 2   Officialname            12394 non-null  object 
 3   Registration_number     12343 non-null  object 
 4   Keph level              12394 non-null  object 
 5   Facility type           12394 non-null  object 
 6   Facility_type_category  12394 non-null  object 
 7   Owner                   12394 non-null  object 
 8   Owner type              12394 non-null  object 
 9   Regulatory body         12394 non-null  object 
 10  Beds                    12394 non-null  object 
 11  Cots                    12394 non-null  int64  
 12  Beds and Cots           12394 non-null  int64  
 13  County                  12394 non-null  object 
 14  Constituency            12394 non-null

In [10]:
# count number of unique names per each Keph level
hospital.groupby('Keph level')['Name'].nunique()

Keph level
Level 2    9411
Level 3    2139
Level 4     818
Level 5      19
Level 6       6
None          1
Name: Name, dtype: int64

Keph level
Level 2    9411

Level 3    2139

Level 4     818

Level 5      19

Level 6       6

None          1

In [11]:
# count number of unique names per each owner type
hospital.groupby('Owner type')['Name'].nunique()

Owner type
Faith Based Organization          1032
Ministry of Health                5878
Non-Governmental Organizations     366
Private Practice                  5118
Name: Name, dtype: int64

Owner type

Faith Based Organization          1032

Ministry of Health                5878

Non-Governmental Organizations     366

Private Practice                  5118

In [12]:
# counting the number of 'DISPENSARY facilities per Owner type 'Faith Based Organization'
hospital[(hospital['Facility type'] == 'DISPENSARY') & (hospital['Owner type'] == 'Faith Based Organization')]['Name'].nunique()

0

The count is zero

In [20]:
# From hospital, print code names for Keph level 5 facilities that have 'Open_weekends' as 'Yes' and 'Open_public_holidays' as 'Yes'
hospital[(hospital['Keph level'] == 5) & (hospital['Open_weekends'] == 'Yes') & (hospital['Open_public_holidays'] == 'Yes')]['Code'].unique()


array([], dtype=object)

In [13]:
import numpy as np

# Use columns 'Open_whole_day', 'Open_public_holidays', 'Open_weekends', 'Open_late_night' to create a new column 'Always_open'
# if all the above columns are 'Yes' then 'Always_open' is 'Yes' else 'No'
hospital['Always_open'] = np.where((hospital['Open_whole_day'] == 'Yes') & (hospital['Open_public_holidays'] == 'Yes') & (hospital['Open_weekends'] == 'Yes') & (hospital['Open_late_night'] == 'Yes'), 'Yes', 'No')

In [15]:
# filter hospital on Always_open == 'Yes'
df = hospital[hospital['Always_open'] == 'Yes']
df.head()

Code                             Name                     Officialname  \
6    None            Genvis Medical Center            Genvis Medical Center   
7   25702                   Annie Hospital                   Annie Hospital   
8   25719       The white cresent Hospital       The white cresent Hospital   
10   None                            C.M.C            Charis Medical Centre   
12   None  PONAH HEALTHCARE MEDICAL CLINIC  PONAH HEALTHCARE MEDICAL CLINIC   

   Registration_number Keph level             Facility type  \
6               016692    Level 3            Medical Center   
7               001654    Level 3            Medical Center   
8                 None    Level 5  Secondary care hospitals   
10                None    Level 3            Medical Clinic   
12               17076    Level 2            Medical Clinic   

   Facility_type_category                                         Owner  \
6          MEDICAL CENTER  Private Practice Lab Technician/Technologist   
7          MEDICAL CENTER            Private Practice - Private Company   
8               HOSPITALS       Private Practice - General Practitioner   
10         MEDICAL CLINIC           Private Practice - Clinical Officer   
12         MEDICAL CLINIC           Private Practice - Clinical Officer   

          Owner type                                    Regulatory body  ...  \
6   Private Practice                                         Kenya MPDB  ...   
7   Private Practice                                         Kenya MPDB  ...   
8   Private Practice                                               None  ...   
10  Private Practice                           Pharmacy & Poisons Board  ...   
12  Private Practice  Kenya Medical Laboratory, Tech & Technologists...  ...   

   Operation status  Open_whole_day  Open_public_holidays Open_weekends  \
6       Operational             Yes                   Yes           Yes   
7       Operational             Yes                   Yes           Yes   
8       Operational             Yes                   Yes           Yes   
10      Operational             Yes                   Yes           Yes   
12      Operational             Yes                   Yes           Yes   

   Open_late_night Service_names Approved Public visible Closed Always_open  
6              Yes           NaN      Yes            Yes     No         Yes  
7              Yes           NaN      Yes            Yes     No         Yes  
8              Yes           NaN      Yes            Yes     No         Yes  
10             Yes           NaN      Yes            Yes     No         Yes  
12             Yes           NaN      Yes            Yes     No         Yes  

[5 rows x 27 columns]

In [ ]:
import matplotlib.pyplot as plt

# plot of Bed count for all level 6 hospitals
hospital[hospital['Keph level'] == 'Level 6'].plot(x='Name', y='Bed capacity', kind='bar', figsize=(10, 6))
plt.title('Bed count for all level 6 hospitals')
plt.xlabel('Hospital Name')
plt.ylabel('Bed count')
plt.show()

